lib/libvfio-user.c

```diff
@@ -506,24 +505,25 @@ handle_device_get_info(vfu_ctx_t *vfu_ctx, uint32_t size,
  * @vfu_ctx: LM context
  * @size: size, in bytes, of the memory pointed to be @dma_regions
  * @map: whether this is a DMA map operation
- * @fds: array of file descriptors. It's length must equal the number of DMA
-         regions, irrespectively if @nr_fds is 0.
- * @nr_fds: size of above array. It must be either 0 or exactly match
- *          the number of DMA regions in @dma_regions.
+ * @fds: array of file descriptors.
+ * @nr_fds: size of above array.
  * @dma_regions: memory that contains the DMA regions to be mapped/unmapped
  *
- * @returns 0 on success, -errno on failure.
+ * @returns 0 on success, -errno on failure. @nr_fds receives the number of
+ * mappable DMA regions that were successfully added.
  */
 int
 handle_dma_map_or_unmap(vfu_ctx_t *vfu_ctx, uint32_t size, bool map,
-                        int *fds, int nr_fds,
+                        int *fds, size_t *nr_fds,
```

Comment:
IMO let's not abuse nr_fds to return back with successfully added DMA regions, as nr_fds are suppose to be the size of fds array. Maybe have separate argument to return added regions.
In-fact for such situations returning >0 is useful.